client/hotrod-client/src/main/java/org/infinispan/client/hotrod/marshall/MarshallerUtil.java

```diff
@@ -21,7 +25,28 @@ private MarshallerUtil() {}
    public static <T> T bytes2obj(Marshaller marshaller, byte[] bytes) {
       if (bytes == null) return null;
       try {
-         return (T) marshaller.objectFromByteBuffer(bytes);
+         Object ret = marshaller.objectFromByteBuffer(bytes);
+
+         // No extra configuration is required for client when using compatibility mode,
+         // and no different marshaller should be required to deal with standard serialization.
+         // So, if the unmarshalled object is still a byte[], it could be a standard
+         // serialized object, so try to deserialise it.
+
+         // NOTE: This has a slight impact on Remote Cache use cases storing
+         // plain byte arrays in cache when compatibility is disabled. Hence,
+         // it might make sense to have a configuration flag in Remote Cache Manager
+         // to make the client behave in compatibility mode, and hence only
+         // try using standard serialization if compatibility is enabledd.
+         if (ret instanceof byte[]) {
+            try (ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream((byte[]) ret))) {
```

Comment:
Maybe a quick check for the "magic" value at the beginning before attempting to read it via OIS ?